### ISSUE TITLE
Print-debugging jax?

### ISSUE THREAD

**Issue by johnsalmon** [google/jax#4615]
I've got some jax code that relies on grad, jvp, etc.  I'd like to do some simple print-style debugging.  E.g., some lines like:

    print("JUtotal: ", ret)
    return ret

In a function whose gradient is required.  I'd really like to see something like:

`JUtotal: -0.924010929514282`

but instead, I see:

```
JUtotal:  Traced<ConcreteArray(-0.9240109029514282)>with<JVPTrace(level=6/0)>
  with primal = Traced<ConcreteArray(-0.9240109029514282)>with<JVPTrace(level=4/0)>
                  with primal = DeviceArray(-0.9240109, dtype=float64)
                       tangent = Traced<ShapedArray(float64[])>with<BatchTrace(level=3/0)>
                                   with val = DeviceArray([ 4.31834536e-04, -5.96999899e-06,  1.25216808e-01,
                                                           -1.07182239e-01, -7.46824003e-07, -8.62923763e-07],            dtype=float64)
                                        batch_dim = 0
       tangent = Traced<ShapedArray(float64[]):JaxprTrace(level=5/0)>
```

I understand why the latter is sometimes important (e.g., when debugging the autograd machinery itself).  But for now, I just want to know the numerical value of 'ret'.  Is there a way to obtain just the numerical value of an object in a traced function with all the extra "hair" removed?

**Comment by shoyer** [comment]
I recommend using `jax.experimental.host_callback` for this -- see `id_print()` in particular: https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html

**Comment by mattjj** [comment]
It fills my heart with joy to see you're still playing with JAX, @johnsalmon !

`id_print` is one option, though its primary purpose is being able to perform prints from code that was staged out to XLA via `jit` (or `pmap`, or control flow primitives). To reach that goal it might not be as ergonomic as Python printing, like you might want in interactive pdb sessions.

The text you see right now is from [the `Tracer.__repr__` method defined in core.py](https://github.com/google/jax/blob/7e801cae3a35d9fb76705f0c23aa715dc6ab9b77/jax/core.py#L538-L544), which is meant to recursively dissect a Tracer and print everything it finds. But as you say, maybe that can be too much information for some common use cases.

One thing we could do here, which @hawkinsp proposed in a chat just now, is make a `Tracer.__str__` that differs from `Tracer.__repr__`, and have the former be more concise. (I believe by only defining `Tracer.__repr__` as we do now, Python reuses that definition for `Tracer.__str__` too.) Then `print(x)` might do what you want, whereas `print(repr(x))` would give more complete information. Maybe I'll give that a shot... 

Any thoughts?

**Comment by shoyer** [comment]
Does it work to write something like `ret.primal`?

I would be reluctant to drop all the extras from the default display, though the detailed type & trace level information could be omitted from `__str__`.

**Comment by mattjj** [comment]
@shoyer by "default display", do you mean what we get from `__repr__` or `__str__`? I agree we need to make both fully-detailed and simplified versions readily accessible.

I think using `ret.primal` is worse because it's violating our Tracer abstractions, whereas using `__repr__` and `__str__` is just calling into a standard API.

**Comment by shoyer** [comment]
> @shoyer by "default display", do you mean what we get from `__repr__` or `__str__`? I agree we need to make both fully-detailed and simplified versions readily accessible.

I was thinking for `__str__` as well.

I guess my concern is that when you're printing tracer, the primal value isn't necessarily any more fundamental than the tangent. For example, inside `linear_transpose()`, the primal value is just a placeholder and the tangent/cotangent is all that matters.

**Comment by mattjj** [comment]
Feedback on this? The called function first calls `print(x)` to show the new short result of `Tracer.__str__`, and then calls `print(repr(x))` to show the more complete info. The code change is in #4618.

```
In [1]: def f(x):
   ...:     print(x)
   ...:     print(repr(x))
   ...:     return x
   ...:

In [2]: from jax import grad
gra
In [3]: grad(f)(3.14)
Traced<3.14>
Traced<ConcreteArray(3.14, weak_type=True)>with<JVPTrace(level=2/0)>
  with primal = Traced<ConcreteArray(3.14, weak_type=True):JaxprTrace(level=1/0)>
       tangent = Traced<ShapedArray(float32[], weak_type=True):JaxprTrace(level=1/0)>
Out[3]: array(1.)

In [4]: from jax import jit

In [5]: jit(f)(3.14)
Traced<float32[]>
Traced<ShapedArray(float32[], weak_type=True)>with<DynamicJaxprTrace(level=0/1)>
Out[5]: DeviceArray(3.14, dtype=float32)

In [6]: grad(grad(f))(3.14)
Traced<3.14>
Traced<ConcreteArray(3.14, weak_type=True)>with<JVPTrace(level=4/0)>
  with primal = Traced<ConcreteArray(3.14, weak_type=True):JaxprTrace(level=3/0)>
       tangent = Traced<ShapedArray(float32[], weak_type=True):JaxprTrace(level=3/0)>
Out[6]: array(0., dtype=float32)
```

I suspect this goes too far for @shoyer but I think it might be worth trying out. It's easy to revise, after all!

**Comment by shoyer** [comment]
This is indeed further than I would go, but we can try it out.

What about `JVPTrace<3.14>` rather than just `Traced<3.14>`?

**Comment by sharadmv** [comment]
Closing because we now have [`jax.debug.print`](https://jax.readthedocs.io/en/latest/debugging/print_breakpoint.html).
